lib/pharos/phase_manager.rb

```diff
@@ -30,12 +30,23 @@ def initialize(**options)
     # @param phases [Array<Pharos::Phases::Base>]
     # @return [Array<...>]
     def run_parallel(phases, &block)
-      threads = phases.map { |phase|
+      threads = phases.map do |phase|
         Thread.new do
           Thread.current.report_on_exception = false
           Retry.perform(yield_object: phase, logger: logger, exceptions: RETRY_ERRORS, &block)
         end
-      }
+      end
+
+      until threads.none?(&:alive?)
+        # Thread status is false when terminated normally, nil when it terminated with exception
+        thread_with_exception = threads.find { |thr| thr.status.nil? }
+        if thread_with_exception
+          threads.map(&:kill) # kill other threads
```

Comment:
I don't think this is safe assumption.

Reply:
Actually it works like it has been working.. now that it's written out it sounds bad.. should we wait all threads to finish before halting whole process?

Reply:
> I don't think this is safe assumption.

Hmm, what is not safe assumption?  `Thread status is false when terminated normally, nil when it terminated with exception` this is fact from the [docs](https://devdocs.io/ruby~2.5/thread#method-i-status) 

> Actually it works like it has been working.. now that it's written out it sounds bad.. should we wait all threads to finish before halting whole process?

Maybe it would be better.

It will delay aborting but shouldn't leave the hosts in a funny state. Maybe larger phases should be split into smaller phases that would complete (and fail) faster.

There was some reason originally to add `Thread.current.abort_on_exception = true` to the pro configure_master

Reply:
> Hmm, what is not safe assumption? 

Sorry for not being too clear.. I meant that killing other threads sounds a bit unsafe.

Reply:
Probably the best way to implement this is to gather all exceptions, report them (with or without backtrace) and then raise a "Phase foobar failed" exception. This way we don't need to brutally kill any threads and user can see errors from every parallel phase.

Reply:
Yeah and that's what happens with `abort_on_exception=true`, and that also happens when you do `threads.map(&:value)`:

```ruby
threads = []
threads << Thread.new { sleep 2; raise "what" }
threads << Thread.new { sleep 5; puts "hello 1" } # <-- will never execute
threads.map(&:value) # because the first `thread.value` call will block + 
                     # reraise, which kills the main thread, which kills
                     # the other threads
```

So we can not call `threads.map(&:join) or `threads.map(&:value)` because both will raise when an exception appears in the thread being waited on.

I think the "let it finish` version would be:

```
Thread.pass until threads.none?(&:alive?)
threads.map(&:value) # this will raise if any of the threads had an exception.
```